### PR TITLE
F v2 extend

### PR DESCRIPTION
**Pull request by ladovina**
Add Covalenthq API + some refactoring.

### CODE REVIEW

File: blockapi/v2/api/covalenth/base.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    supported_requests = {
+        'get_balance': '/v1/{chain_id}/address/{address}/balances_v2/'
+    }
+
+    def __init__(self, chain_id: int, api_key: str):
+        super().__init__(api_key)
+        self.chain_id = chain_id
+
+        # Set http basic auth for requests.
+        self._session.auth = (api_key, "")
```
Comment (galvanizze) [inline blockapi/v2/api/covalenth/base.py:31]: nápad: 

`CovalentApiBase` by mohla byť abstract class a mať definovaný `chain_id = None` (btw rovnako ako to máš aj pre `api_options`). Tým pádom by `__init__` nemusel mať argument `chain_id` a mohol by si ho vymazať z každej child classy, ktorá dedí od `CovalentApiBase`.

File: blockapi/v2/api/covalenth/base.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        return self._parse_items(response)
+
+    def _parse_items(self, response: Dict) -> Optional[BalanceItem]:
+        try:
+            raw_balances = response['data']['items']
+        except KeyError as e:
+            logger.exception(e)
+            return []
+
+        for raw_balance in raw_balances:
```
Comment (galvanizze) [inline blockapi/v2/api/covalenth/base.py:45]: atribút `self.coin` (základná mena blockchainu) som chcel mať na úrovni tých class definovanú, aby sme pre viac rôznych API providerov nemuseli definovať ten coin znovu a aj aby to nemuselo byť vždy definované v kóde.

Covalent to má však trošku iný interface ako ostatné API služby, base balance sa vracia medzi tokenmi, z dokumentácie: "As a special case, network native tokens like ETH on Ethereum are also returned even though it's not a token contract.". Ale i tak by som chcel aby sme vždy ošetrili tento prípad a definovali konštantu `self.coin` s ručne vyplnenými parametrami, mimo iné i `coingecko_id` je pre nás veľmi dôležitá info a zrýchľuje mapovanie v Crypkite.

Tu je príklad base currency (ETH) v response:

image

File: blockapi/test/e2e/v2/test_covalent.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        (HECOCovalentApi, "0xfc43f5f9dd45258b3aff31bdbe6561d97e8b71de"),
+        (IoTEXCovalentApi, "0xfc43f5f9dd45258b3aff31bdbe6561d97e8b71de"),
+        (KlaytnCovalentApi, "0xfc43f5f9dd45258b3aff31bdbe6561d97e8b71de"),
+        (MoonBeamCovalentApi, "0xfc43f5f9dd45258b3aff31bdbe6561d97e8b71de"),
+        (PalmCovalentApi, "0xfc43f5f9dd45258b3aff31bdbe6561d97e8b71de"),
+        (MaticCovalentApi, "0xfc43f5f9dd45258b3aff31bdbe6561d97e8b71de"),
+        (RskCovalentApi, "0xfc43f5f9dd45258b3aff31bdbe6561d97e8b71de"),
+    ],
+)
+def test_get_balances(api_class: Type[CovalentApiBase], address: str) -> None:
```
Comment (galvanizze) [inline blockapi/test/e2e/v2/test_covalent.py:46]: tento základný get_balance testuje aj generický test; aktuálne tam nie je definovaný ten rate limit, ale radšej by som to doplnil do toho generic testu. čo myslíš?

File: blockapi/test/v2/api/covalenth/test_ethereum.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    with open(json_path) as f:
+        return json.load(f)
+
+
+@pytest.fixture
+def eth_covalent_api():
+    return EthCovalentApi(api_key="test_key")
+
+
+def test_parse_balance(eth_covalent_api, covalenth_eth_balances_response):
```
Comment (galvanizze) [inline blockapi/test/v2/api/covalenth/test_ethereum.py:24]: ja na takýto typ testov, keď potrebujem response používam VCR mock, poznáš?

[príklad z v1](https://github.com/crypkit/blockapi/blob/master/blockapi/test/api/test_solana.py#L12)

File: blockapi/v2/api/covalenth/avalanche.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+from blockapi.v2.base import ApiOptions
+from blockapi.v2.models import Blockchain
+
+
+class AvalancheCovalentApi(CovalentApiBase):
+
+    CHAIN_ID = 43114
+
+    api_options = ApiOptions(
+        blockchain=Blockchain.ETHEREUM,
```
Comment (galvanizze) [inline blockapi/v2/api/covalenth/avalanche.py:11]: zlý chain

File: blockapi/v2/api/covalenth/polygon.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+from blockapi.v2.base import ApiOptions
+from blockapi.v2.models import Blockchain
+
+
+class MaticCovalentApi(CovalentApiBase):
+
+    CHAIN_ID = 137
+
+    api_options = ApiOptions(
+        blockchain=Blockchain.ETHEREUM,
```
Comment (galvanizze) [inline blockapi/v2/api/covalenth/polygon.py:11]: zlý chain

File: blockapi/v2/base.py [hunk 7]
```diff
@@ -71,10 +75,15 @@ def _opt_raise_on_other_error(self, response: Response) -> None:
     def __repr__(self):
         return (
             f"{self.__class__.__name__}"
-            f"(coin={self.coin.name},address={self.address})"
+            f"(coin={self.coin.name})"
         )
 
 
+class IBalance(ABC):
+    def get_balance(self, *args, **kwargs) -> List[BalanceItem]:
```
Comment (galvanizze) [inline blockapi/v2/base.py:83]: tu môžeš pridať argument `address`, ten je povinný

**Review by galvanizze — CHANGES_REQUESTED** [review]
mám len pár poznámok, inak fajn PR 👍 :)

File: blockapi/v2/api/covalenth/base.py (hunk 1 above)
Comment (ladovina) [reply blockapi/v2/api/covalenth/base.py:31]: Yes I was thinking about it - but current solution has one big plus: we force child classes to set chain_id which I like.

Currently `api_options` is not forced to be implemented which I do not like - there are ways how to do it even with class attr. but to have it in constructor like `chain_id` is simple which I probably prefer...

Comment (ladovina) [reply blockapi/v2/api/covalenth/base.py:31]: Changed to abstractmethod + property.

File: blockapi/v2/api/covalenth/base.py (hunk 2 above)
Comment (ladovina) [reply blockapi/v2/api/covalenth/base.py:58]: self.coin added + added skipping of "blockchain-native" coins in "_parse_items".

File: blockapi/v2/api/covalenth/avalanche.py (hunk 5 above)
Comment (ladovina) [reply blockapi/v2/api/covalenth/avalanche.py]: fixed

File: blockapi/v2/api/covalenth/polygon.py (hunk 6 above)
Comment (ladovina) [reply blockapi/v2/api/covalenth/polygon.py]: fixed

File: blockapi/v2/base.py (hunk 7 above)
Comment (ladovina) [reply blockapi/v2/base.py:83]: I am for strictly define it as `get_balance(self, address:str)` or leave it as it is. 

That means I do not want to mix it to sth like this: `get_balance(self, address, *args, **kwargs)` what do you think?

Comment (galvanizze) [reply blockapi/v2/base.py:83]: maybe we can include at least `**kwargs` as backdoors for special attributes? Or leave it just with address as you proposed.

File: blockapi/test/e2e/v2/test_covalent.py (hunk 3 above)
Comment (ladovina) [reply blockapi/test/e2e/v2/test_covalent.py:46]: yes I will move it there :+1:

File: blockapi/test/v2/test_data.py [hunk 8]
```diff
@@ -2,10 +2,10 @@
 
 from blockapi.v2.api.ethplorer import EthplorerApi
 from blockapi.v2.api.solana import SolanaApi
-from blockapi.v2.api.terra_money import TerraMoneyApi
+from blockapi.v2.api.terra import TerraApi
 
 # TODO create method for auto loading all classes
-API_CLASSES = [EthplorerApi, SolanaApi, TerraMoneyApi]
+API_CLASSES = [EthplorerApi, SolanaApi, TerraApi]
```
[the remark this answers is not in the document]
Comment (ladovina) [reply blockapi/test/v2/test_data.py]: added

File: blockapi/v2/base.py (hunk 7 above)
Comment (ladovina) [reply blockapi/v2/base.py:83]: changed to strict "get_balance(self, address: str)"

File: blockapi/test/v2/api/covalenth/test_ethereum.py (hunk 4 above)
Comment (ladovina) [reply blockapi/test/v2/api/covalenth/test_ethereum.py:24]: Ok, I didn't know this.

But if you do not insist on it i would stick to simple json fixture - simplicity, transparent, no 3rd party deps + I want to test only parsing here (no http which is tested in test general).

File: blockapi/v2/api/covalenth/astar.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    api_options = ApiOptions(
+        blockchain=Blockchain.ASTAR,
+        base_url=CovalentApiBase.API_BASE_URL,
+        rate_limit=CovalentApiBase.API_BASE_RATE_LIMIT,
+    )
+
+    coin = COIN_SDN
+
+    def __init__(self, api_key: str):
+        super().__init__(api_key)
```
Comment (galvanizze) [inline blockapi/v2/api/covalenth/astar.py:20]: if there is no change in `__init__` (comparing to parent), you can omit it in child class

**Review by galvanizze — CHANGES_REQUESTED** [review]
very nice, just one thing, thank you :)

File: blockapi/v2/api/covalenth/astar.py (hunk 9 above)
Comment (ladovina) [reply blockapi/v2/api/covalenth/astar.py:20]: good point, I will change it.

Comment (ladovina) [reply blockapi/v2/api/covalenth/astar.py]: changed

File: blockapi/test/v2/api/covalenth/test_ethereum.py (hunk 4 above)
Comment (galvanizze) [reply blockapi/test/v2/api/covalenth/test_ethereum.py:24]: makes sense

**Pull request merged**